lib/dockerregistry/inventory.go

```diff
@@ -872,6 +880,63 @@ func getRegistryTagsFrom(req stream.ExternalRequest) (*google.Tags, error) {
 	return tags, nil
 }
 
+func getGManifestListWrapper(
+	req stream.ExternalRequest) (*GManifestList, error) {
+
+	var gManifestList *GManifestList
+
+	var getGManifestListCondition wait.ConditionFunc = func() (bool, error) {
+		var err error
+
+		gManifestList, err = getGManifestListFrom(req)
+
+		// We never return an error (err) in the second part of our return
+		// argument, because we don't want to prematurely stop the
+		// ExponentialBackoff() loop; we want it to continue looping until
+		// either we get a well-formed value, or until it hits
+		// ErrWaitTimeout. This is how ExponentialBackoff() uses the
+		// ConditionFunc type.
+		if err == nil &&
+			gManifestList != nil &&
+			len(gManifestList.Manifests) > 0 {
+
+			return true, nil
+		}
+		return false, nil
+	}
+
+	err := wait.ExponentialBackoff(
+		stream.BackoffDefault,
+		getGManifestListCondition)
+
+	if err != nil {
+		klog.Error(err)
+		return nil, err
+	}
+
+	return gManifestList, nil
+}
+
+func getGManifestListFrom(req stream.ExternalRequest) (*GManifestList, error) {
+	reader, _, err := req.StreamProducer.Produce()
+	if err != nil {
+		klog.Warning("error reading from stream:", err)
```

Comment:
Ah - you're logging it here.  It's typically better to return the error and let the caller choose to log or not, IMO

Reply:
lol Github hid your comment here so I didn't see it until l unfolded this convo... Anyway ya I will change the logging convention to make the caller log it (it feels cleaner that way to me actually, dunno why I didn't do it that way the first time)

Reply:
Oh wait, I just remembered. It's because of the way ExponentialBackoff behaves. If I return an error here, then the backoff loop will exit (on instead of letting the backoff mechanism exit the backoff loop due to timeout). See https://godoc.org/k8s.io/apimachinery/pkg/util/wait#ConditionFunc

That explains the non-idiomatic code... still, I can add logging to make the various different error cases more explicit.